sdk/template/Azure.Template/src/Generated/MultiVersionClient.cs

```diff
@@ -0,0 +1,237 @@
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        /// <param name="context"> The request context, which can override default behaviors of the client pipeline on a per-call basis. </param>
+        /// <exception cref="RequestFailedException"> Service returned a non-success status code. </exception>
+        /// <returns> The response returned from the service. </returns>
+        /// <example>
+        /// This sample shows how to call ServiceV1Operation and parse the result.
+        /// <code><![CDATA[
+        /// var credential = new DefaultAzureCredential();
+        /// var endpoint = new Uri("<https://my-service.azure.com>");
+        /// var client = new MultiVersionClient(endpoint, credential);
+        /// 
+        /// Response response = client.ServiceV1Operation();
+        /// 
+        /// JsonElement result = JsonDocument.Parse(response.ContentStream).RootElement;
+        /// Console.WriteLine(result.ToString());
+        /// ]]></code>
+        /// </example>
+        public virtual Response ServiceV1Operation(RequestContext context = null)
+        {
+            using var scope = ClientDiagnostics.CreateScope("MultiVersionClient.ServiceV1Operation");
+            scope.Start();
+            try
+            {
+                using HttpMessage message = CreateServiceV1OperationRequest(context);
+                return _pipeline.ProcessMessage(message, context);
+            }
+            catch (Exception e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+        }
+
+        /// <summary>Servicev2Operation method. Added in service version 2022-02-02. </summary>
+        /// <param name="context"> The request context, which can override default behaviors of the client pipeline on a per-call basis. </param>
+        /// <exception cref="RequestFailedException"> Service returned a non-success status code. </exception>
+        /// <returns> The response returned from the service. </returns>
+        /// <example>
+        /// This sample shows how to call ServiceV2OperationAsync and parse the result.
+        /// <code><![CDATA[
+        /// var credential = new DefaultAzureCredential();
+        /// var endpoint = new Uri("<https://my-service.azure.com>");
+        /// var client = new MultiVersionClient(endpoint, credential);
+        /// 
+        /// Response response = await client.ServiceV2OperationAsync();
+        /// 
+        /// JsonElement result = JsonDocument.Parse(response.ContentStream).RootElement;
+        /// Console.WriteLine(result.ToString());
+        /// ]]></code>
+        /// </example>
+        public virtual async Task<Response> ServiceV2OperationAsync(RequestContext context = null)
+        {
+            // Generator knows from Cadl @added decorator which version an operation was added in.
+            if (_serviceVersion < MultiVersionClientOptions.ServiceVersion.V2022_02_02)
```

Comment:
Throw NotSupportedException if calling from client created for service version that does not support the operation.

Reply:
Given a recent DPG meeting, I did want to point out that - even if we relegate errors to the service for model properties - I don't think we can/should for endpoints. Some services, like Key Vault, return a 404 along with the default ASP.NET error *HTML* page for unmapped endpoints, which is less than ideal and can be confusing. And if we're validating service versions for some things, why not consider it for everything? Why incur network IO if we know something won't work or, worse, work incorrectly if a service doesn't handle PATCH-like semantics correctly.

/cc @KrzysztofCwalina @tg-msft